src/consensus/upgrades.h

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2018 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef ZCASH_CONSENSUS_UPGRADES_H
+#define ZCASH_CONSENSUS_UPGRADES_H
+
+#include "consensus/params.h"
+
+#include <boost/optional.hpp>
+
+enum UpgradeState {
+    UPGRADE_DISABLED,
+    UPGRADE_PENDING,
+    UPGRADE_ACTIVE
+};
+
+struct NUInfo {
+    /** Branch ID (a random non-zero 32-bit value) */
+    uint32_t nBranchId;
+    /** User-facing name for the upgrade */
+    std::string strName;
+    /** User-facing information string about the upgrade */
+    std::string strInfo;
+};
+
+extern const struct NUInfo NetworkUpgradeInfo[];
+
+/**
+ * Checks the state of a given network upgrade based on block height.
+ * Caller must check that the height is >= 0 (and handle unknown heights).
+ */
+UpgradeState NetworkUpgradeState(
+    int nHeight,
+    const Consensus::Params& params,
+    Consensus::UpgradeIndex idx);
+
+/**
+ * Returns true if the given network upgrade is active as of the given block
+ * height. Caller must check that the height is >= 0 (and handle unknown
+ * heights).
```

Comment:
A negative height value usually means the height is unknown, but it could also be that the caller doesn't actually know the height at this point. In the latter case, the caller can't use this function, and needs to decide what to do itself.